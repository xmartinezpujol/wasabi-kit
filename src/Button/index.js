import React from 'react';
import styled from 'styled-components';

import COLOR_PALETTE from '../Constants';

const shape = {
  round: '200px',
  square: 0,
  default: 6
};

const size = {
  small: {
    fontSize: 14
  },
  medium: {
    fontSize: 18
  },
  large: {
    fontSize: 22
  }
};

const template = props => {
  const templateRemake = {
    default: {
      background: COLOR_PALETTE[props.type],
      border: 0,
      color: COLOR_PALETTE[props.color]
    },
    dynamic: {
      color: COLOR_PALETTE.purewhite,
      position: 'relative',
      zIndex: 1,
      border: 0,
      background: `linear-gradient(to right, ${props.gradient[0]}, ${props.gradient[1]})`,
      '&:after': {
        content: "''",
        bottom: '-10px',
        left: '5%',
        height: '100%',
        width: '90%',
        position: 'absolute',
        background: 'inherit',
        top: '0.5rem',
        filter: 'blur(0.4rem)',
        opacity: 0.65,
        zIndex: -1
      }
    },
    outlined: {
      background: COLOR_PALETTE[props.type],
      border: '2px solid',
      borderColor: COLOR_PALETTE[props.color],
      color: COLOR_PALETTE[props.color],
      ':hover:enabled': {
        background: COLOR_PALETTE[props.color],
        color: COLOR_PALETTE[props.type]
      }
    },
    link: {
      backgroundColor: props.type ? COLOR_PALETTE[props.type] : COLOR_PALETTE.transparent,
      border: 0,
      color: COLOR_PALETTE[props.color],
      ':hover:enabled': {
        boxShadow: 'none',
        backgroundColor:
          props.type === COLOR_PALETTE.transparent
            ? COLOR_PALETTE.transparent
            : COLOR_PALETTE[props.color],
        color:
          props.type !== COLOR_PALETTE.transparent
            ? COLOR_PALETTE[props.type]
            : COLOR_PALETTE[props.color]
      }
    }
  };

  const theme = templateRemake[props.template ? props.template : 'default'];
  const templateMod = Object.assign({}, theme);

  return templateMod;
};

const ButtonBox = styled.button(
  {
    fontFamily: 'Poppins, sans-serif',
    cursor: 'pointer',
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'center',
    transition: '0.25s cubic-bezier(0.17, 0.67, 0.52, 0.97)',
    WebkitTapHighlightColor: 'rgba(255, 255, 255, 0)',
    ':hover:enabled': {
      boxShadow: '0 7px 14px rgba(50,50,93,.1), 0 3px 6px rgba(0,0,0,.08)'
    },
    ':focus': {
      outline: 0
    },
    ':disabled': {
      opacity: 0.5,
      cursor: 'not-allowed'
    }
  },
  props => ({
    ...template(props),
    ...size[props.size],
    borderRadius: shape[props.shape],
    fontWeight: props.bold ? 700 : 500,
    padding: props.padding,
    minHeight: props.minHeight,
    minWidth: props.minWidth,
    textTransform: props.uppercase ? 'uppercase' : 'initial',
    transform: props.noanim ? 'none' : 'translateY(-1px)',
    ':active:enabled': {
      transform: props.noanim ? 'none' : 'translateY(1px)'
    },
    ':hover': {
      transform: props.noanim ? 'none' : 'translateY(-5px)'
    },
    ':hover:after': {
      bottom: props.noanim ? 'inherit' : '-10px'
    },
    ':not(:last-child)': {
      margin: props.margin
    }
  })
);

const Button = props => (
  <ButtonBox data-cy="button" {...props}>
    {props.children}
  </ButtonBox>
);

Button.defaultProps = {
  type: 'wasabi',
  template: 'default',
  color: 'nori',
  bold: false,
  shape: 'default',
  size: 'medium',
  minHeight: '50px',
  minWidth: '50px',
  noanim: false,
  uppercase: false,
  gradient: ['#0fd850', '#35FF69'],
  padding: '10px 35px',
  margin: '0 10px 0 0'
};

export default Button;
