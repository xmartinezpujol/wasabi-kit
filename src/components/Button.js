import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import COLOR_PALETTE from '../Constants';

const shape = {
  round: '200px',
  square: 0,
  default: 6,
};

const size = {
  small: {
    fontSize: 14,
    minHeight: 30,
  },
  medium: {
    fontSize: 18,
    minHeight: 54,
  },
  large: {
    fontSize: 22,
    minHeight: 60,
  },
};

const template = (props) => {
  const templateRemake = {
    default: {
      background: COLOR_PALETTE[props.type],
      border: 0,
      color: COLOR_PALETTE.nori,
    },
    outlined: {
      backgroundColor: COLOR_PALETTE.rice,
      border: '2px solid',
      borderColor: COLOR_PALETTE[props.bordercolor],
      color: props.bordercolor
        ? COLOR_PALETTE[props.bordercolor]
        : COLOR_PALETTE.night,
      ':hover:enabled': {
        background: (props.bordercolor === COLOR_PALETTE.nori || !props.bordercolor)
          ? COLOR_PALETTE.night
          : COLOR_PALETTE.rice,
        color: (props.bordercolor === COLOR_PALETTE.nori || !props.bordercolor)
          ? COLOR_PALETTE.rice
          : props.bordercolor,
      },
    },
    link: {
      backgroundColor: COLOR_PALETTE[props.type],
      border: 0,
      color: props.bordercolor
        ? COLOR_PALETTE[props.bordercolor]
        : COLOR_PALETTE.night,
      ':hover:enabled': {
        boxShadow: 'none',
        color: props.bordercolor
          ? props.bordercolor === COLOR_PALETTE.nori
            ? COLOR_PALETTE.rice
            : props.bordercolor
          : COLOR_PALETTE.nori,
      },
    },
  };

  const theme = templateRemake[props.template ? props.template : 'default'];
  const templateMod = Object.assign(
    {},
    theme,
  );

  return templateMod;
};

const Icon = styled.i(
  {
    display: 'flex',
  },
  props => ({
    fontSize: props.iconSize,
    width: props.iconWidth,
    fontWeight: props.iconWeight,
  }),
);

const ButtonBox = styled.button(
  {
    fontFamily: 'Poppins, sans-serif',
    margin: '0 10px',
    cursor: 'pointer',
    display: 'flex',
    alignItems: 'center',
    transition: '0.25s cubic-bezier(0.17, 0.67, 0.52, 0.97)',
    padding: '0 22px',
    WebkitTapHighlightColor: 'rgba(255, 255, 255, 0)',
    ':hover:enabled': {
      boxShadow: '0 7px 14px rgba(50,50,93,.1), 0 3px 6px rgba(0,0,0,.08)',
    },
    ':focus': {
      outline: 0,
    },
    ':disabled': {
      opacity: 0.5,
      cursor: 'not-allowed',
    },
  },
  props => ({
    ...template(props),
    ...size[props.size],
    borderRadius: shape[props.shape],
    fontWeight: props.bold ? 700 : 500,
    minWidth: props.fullWidth ? '100%' : 'initial',
    textTransform: props.uppercase ? 'uppercase' : 'initial',
    transform: props.noanim ? 'none' : 'translateY(-1px)',
    ':active:enabled': {
      transform: props.noanim ? 'none' : 'translateY(1px)',
    },
  }),
);

const Button = props => (
  <ButtonBox {...props}>
    {props.icon !== ''
    && (
      <Icon
        iconWeight={props.iconWeight}
        iconSize={props.iconSize}
        iconWidth={props.iconWidth}
        className={`${props.iFont} ${props.iFont}-${props.icon}`}
        aria-hidden="true"
      />
    )}
    {props.children}
  </ButtonBox>
);

Button.propTypes = {
  type: PropTypes.string,
  template: PropTypes.string,
  color: PropTypes.string,
  bordercolor: PropTypes.string,
  bold: PropTypes.bool,
  shape: PropTypes.string,
  size: PropTypes.string,
  icon: PropTypes.string,
  iconSize: PropTypes.number,
  iconWeight: PropTypes.number,
  iconWidth: PropTypes.number,
  iFont: PropTypes.string,
  fullWidth: PropTypes.bool,
  noanim: PropTypes.bool,
  uppercase: PropTypes.bool,
};

Button.defaultProps = {
  type: 'wasabi',
  template: 'default',
  color: 'wasabi',
  bordercolor: COLOR_PALETTE.night,
  bold: false,
  shape: 'default',
  size: 'medium',
  icon: '',
  iconSize: 22,
  iconWeight: 300,
  iconWidth: 22,
  iFont: 'fa',
  fullWidth: false,
  noanim: false,
  uppercase: false,
};

export default Button;
