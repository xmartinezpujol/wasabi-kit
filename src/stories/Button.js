import React from 'react';

import { storiesOf } from '@storybook/react';
import { withKnobs, text, boolean, select, number } from '@storybook/addon-knobs';

import Button from '../Button';
import View from '../View';
import Icon from '../Icon';

const optionsTemplate = {
  default: 'default',
  outlined: 'outlined',
  link: 'link',
  dynamic: 'dynamic'
};
const optionsColor = {
  black: 'black',
  yellow: 'yellow',
  blue: 'blue',
  red: 'red',
  wasabi: 'wasabi',
  sashimi: 'sashimi',
  transparent: 'transparent'
};
storiesOf('Button', module)
  .addDecorator(withKnobs)
  .add('default', () => {
    const defaultTeamplate = 'default';
    const defaultColor = 'black';
    const defaultBackgroundColor = 'wasabi';

    const label = 'Theme';
    const label2 = 'Color';
    const label3 = 'Background';

    const template = select(label, optionsTemplate, defaultTeamplate);
    const color = select(label2, optionsColor, defaultColor);
    const background = select(label3, optionsColor, defaultBackgroundColor);

    return (
      <View
        style={{ width: '100vw', height: '100vh' }}
        container
        align="center"
        justify="center"
        direction="column"
      >
        <Button
          color={color}
          type={background}
          bold={boolean('Bold', true)}
          disabled={boolean('Disabled', false)}
          uppercase={boolean('Uppercase', true)}
          template={template}
          noanim={boolean('No Animation', false)}
        >
          {text('text 1', 'Button')}
        </Button>
      </View>
    );
  })
  .add('with icon', () => {
    const defaultTeamplate = 'default';
    const defaultColor = 'black';
    const defaultIconColor = 'red';
    const defaultBackground = 'wasabi';

    const label = 'Theme';
    const label2 = 'Color';
    const label3 = 'Icon Color';
    const label4 = 'Background Color';

    const template = select(label, optionsTemplate, defaultTeamplate);
    const color = select(label2, optionsColor, defaultColor);
    const iconColor = select(label3, optionsColor, defaultIconColor);
    const background = select(label4, optionsColor, defaultBackground);

    return (
      <View
        style={{ width: '100vw', height: '100vh' }}
        container
        align="center"
        justify="center"
        direction="column"
      >
        <Button
          padding={text('Padding', '0')}
          color={color}
          type={background}
          bold={boolean('Bold', true)}
          disabled={boolean('Disabled', false)}
          uppercase={boolean('Uppercase', true)}
          template={template}
          noanim={boolean('No Animation', false)}
        >
          <View width={150} justify="space-evenly" align="center">
            {text('text 1', 'Button')}
            <Icon
              font={text('Icon Font', 'fa')}
              name={text('Icon name (FA)', 'comment')}
              size={number('Icon fontSize', 22)}
              color={iconColor}
            />
          </View>
        </Button>
      </View>
    );
  })
  .add('with icon only', () => {
    const defaultTeamplate = 'default';
    const defaultborderColor = 'black';
    const defaultColor = 'red';
    const defaultBackgroundColor = 'wasabi';

    const label = 'Theme';
    const label2 = 'Color';
    const label3 = 'Icon Color';
    const label4 = 'Background';

    const template = select(label, optionsTemplate, defaultTeamplate);
    const borderColor = select(label2, optionsColor, defaultborderColor);
    const iconColor = select(label3, optionsColor, defaultColor);
    const background = select(label4, optionsColor, defaultBackgroundColor);

    return (
      <View
        style={{ width: '100vw', height: '100vh' }}
        container
        align="center"
        justify="center"
        direction="column"
      >
        <Button
          padding={text('Padding', '0')}
          shape="round"
          color={borderColor}
          type={background}
          bold={boolean('Bold', true)}
          disabled={boolean('Disabled', false)}
          uppercase={boolean('Uppercase', true)}
          template={template}
          noanim={boolean('No Animation', false)}
        >
          <Icon
            font={text('Icon Font', 'fa')}
            name={text('Icon name (FA)', 'comment')}
            size={number('Icon fontSize', 22)}
            color={iconColor}
          />
        </Button>
      </View>
    );
  });
