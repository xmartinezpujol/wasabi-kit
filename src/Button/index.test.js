import React from 'react';
import { shallow, mount } from 'enzyme';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import Button from '.';
import Icon from '../Icon/';

import COLOR_PALETTE from '../Constants';

describe('Button', () => {
  it('should render a Button with no icon', () => {
    const tree = renderer.create(<Button />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should render a Button with icon', () => {
    const props = {};
    const iconProps = {
      icon: 'bomb',
      iconFont: 'fa',
      iconSize: 20,
      iconWidth: 10
    };
    const tree = renderer.create(<Button {...props} />).toJSON();
    expect(tree).toMatchSnapshot();

    const wrapper = shallow(
      <Button {...props}>
        <Icon {...iconProps} />
        Text
      </Button>
    );
    expect(wrapper.find('Icon').length).toBe(1);
  });
  it('should have prop styles', () => {
    const props = {
      type: 'grey',
      template: 'link',
      color: COLOR_PALETTE.nori,
      bold: true,
      shape: 'square',
      uppercase: true
    };
    const wrapper = mount(<Button {...props} />);
    expect(wrapper).toHaveStyleRule('font-weight', '700');
    expect(wrapper).toHaveStyleRule('text-transform', 'uppercase');
    expect(wrapper).toHaveStyleRule('border-radius', '0');
    const newProps = {
      type: 'green'
    };
    const wrapperButton = mount(<Button {...newProps} />);
    expect(wrapperButton).toHaveStyleRule('background', COLOR_PALETTE.green);
    expect(wrapperButton).toHaveStyleRule('color', COLOR_PALETTE.nori);
  });
});
