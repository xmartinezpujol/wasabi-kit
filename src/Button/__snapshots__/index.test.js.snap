// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render a Button with icon 1`] = `
.c0 {
  font-family: Poppins,sans-serif;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: 0.25s cubic-bezier(0.17,0.67,0.52,0.97);
  transition: 0.25s cubic-bezier(0.17,0.67,0.52,0.97);
  -webkit-tap-highlight-color: rgba(255,255,255,0);
  background: #D4EE5E;
  border: 0;
  color: #424649;
  font-size: 18px;
  border-radius: 6px;
  font-weight: 500;
  padding: 10px 35px;
  min-height: 50px;
  min-width: 50px;
  text-transform: initial;
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
}

.c0:hover:enabled {
  box-shadow: 0 7px 14px rgba(50,50,93,.1),0 3px 6px rgba(0,0,0,.08);
}

.c0:focus {
  outline: 0;
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0:active:enabled {
  -webkit-transform: translateY(1px);
  -ms-transform: translateY(1px);
  transform: translateY(1px);
}

.c0:hover {
  -webkit-transform: translateY(-5px);
  -ms-transform: translateY(-5px);
  transform: translateY(-5px);
}

.c0:hover:after {
  bottom: -10px;
}

.c0:not(:last-child) {
  margin: 0 10px 0 0;
}

<button
  className="c0"
  color="nori"
  data-cy="button"
  shape="default"
  size="medium"
  type="wasabi"
/>
`;

exports[`Button should render a Button with no icon 1`] = `
.c0 {
  font-family: Poppins,sans-serif;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: 0.25s cubic-bezier(0.17,0.67,0.52,0.97);
  transition: 0.25s cubic-bezier(0.17,0.67,0.52,0.97);
  -webkit-tap-highlight-color: rgba(255,255,255,0);
  background: #D4EE5E;
  border: 0;
  color: #424649;
  font-size: 18px;
  border-radius: 6px;
  font-weight: 500;
  padding: 10px 35px;
  min-height: 50px;
  min-width: 50px;
  text-transform: initial;
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
}

.c0:hover:enabled {
  box-shadow: 0 7px 14px rgba(50,50,93,.1),0 3px 6px rgba(0,0,0,.08);
}

.c0:focus {
  outline: 0;
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0:active:enabled {
  -webkit-transform: translateY(1px);
  -ms-transform: translateY(1px);
  transform: translateY(1px);
}

.c0:hover {
  -webkit-transform: translateY(-5px);
  -ms-transform: translateY(-5px);
  transform: translateY(-5px);
}

.c0:hover:after {
  bottom: -10px;
}

.c0:not(:last-child) {
  margin: 0 10px 0 0;
}

<button
  className="c0"
  color="nori"
  data-cy="button"
  shape="default"
  size="medium"
  type="wasabi"
/>
`;
